### PR TITLE
Handle reconnect on begin

### PR DESCRIPTION
**Pull request by liveforeverx**

### CODE REVIEW

File: lib/grakn/channel.ex [hunk 1]
```diff
@@ -14,27 +14,39 @@ defmodule Grakn.Channel do
...
-        open_transaction(channel, tx_request)
-      else
-        error
+  defp may_retry(channel, tx_request, error, cached?) do
+    %{keyspace: keyspace, conn_opts: conn_opts} = tx_request
+
+    with {:error, %GRPC.RPCError{message: message}} when is_binary(message) <- error do
+      cond do
+        cached? and (message =~ ~r/closed/ or message =~ ~r/null\..*/) ->
+          # If session was closed by grakn, so we remove it from cache and try again
```
Comment (lorenzosinisi) [inline lib/grakn/channel.ex]: little typo "so we remove it from the cache and try again"

File: lib/grakn/channel.ex [hunk 2]
```diff
@@ -14,27 +14,39 @@ defmodule Grakn.Channel do
...
+        with {:ok, tx} <- Grakn.Transaction.new(channel, opts),
+             {:ok, tx} <- Transaction.open(tx, session_id, type, opts) do
+          {:ok, channel, tx, session_id}
+        else
+          error ->
+            may_retry(channel, tx_request, error, cached?)
+        end
+
+      error ->
+        may_retry(channel, tx_request, error, false)
```
Comment (lorenzosinisi) [inline lib/grakn/channel.ex]: We could add some tests for this new logic, what do you think?

File: lib/grakn/channel.ex [hunk 3]
```diff
@@ -14,27 +14,39 @@ defmodule Grakn.Channel do
...
 
-  defp may_retry_session(channel, %{keyspace: keyspace} = tx_request, error, cached?) do
-    with {:error, %GRPC.RPCError{message: message}} when cached? and not is_nil(message) <- error do
-      if message =~ ~r/session.*closed/ or message =~ ~r/null\..*/ do
-        # If session was closed by grakn, so we remove it from cache and try again
-        Cache.delete({:keyspace, keyspace})
-        open_transaction(channel, tx_request)
-      else
-        error
+  defp may_retry(channel, tx_request, error, cached?) do
```
Comment (lorenzosinisi) [inline lib/grakn/channel.ex]: naming: what about callingthis func "maybe_reopen_transaction"

File: lib/grakn/channel.ex (hunk 2 above)
Comment (liveforeverx) [reply lib/grakn/channel.ex]: The tests would be hacky, as I need to add API, which expose internals of channel implementation to simulate observed behaviour in production.

What do you think, should we do this?

**Pull request merged**